drivers/spi/spi_xec_gpspi.c

```diff
@@ -0,0 +1,791 @@
+
+	ctx->config = spi_conf;
+
+	memcpy(&data->scfg, spi_conf, sizeof(struct spi_config));
+
+	regs->enable |= BIT(XEC_GPSPI_ENABLE_EN_POS);
+	if (regs->status & BIT(XEC_GPSPI_STATUS_RXBF_POS)) {
+		/* clear RX buffer */
+		regs->control &= ~(BIT(XEC_GPSPI_CTRL_AUTO_READ_POS));
+		regs->status = regs->rx_data;
+		regs->status = regs->rx_data;
+	}
+
+	return 0;
+}
+
+/* Synchronous (blocking) transfer.
+ * The configuration routine programs the controller for full-duplex or dual I/O mode.
+ * GPSPI controller requires a byte write to its TX data register to generate
+ * SPI clocks. The controller always samples input line(s) on the receive clock
+ * edge specified by CPOL/CPHA. Once 8-bits is recevied it is stored in GPSPI
+ * RX data register and RXBF status is set. We must read RX data register to
+ * clear RXBF. If RX data is not read and more clocks are generated then the HW
+ * will require at least to back to back reads of RX data to clear RXBF.
+ * SPI buffer handling:
+ * When transmit buffers are exhausted we set I/O direction to input and write a 0
+ * to GPSPI TX data register to generate SPI clocks. Data sampled on input line(s)
+ * is stored if a receive buffer exists or discarded otherwise.
+ */
+static int xec_gpspi_xfr_sync(const struct device *dev,
+			      const struct spi_config *spi_conf)
+{
+	const struct spi_xec_gpspi_config *cfg = dev->config;
+	struct spi_xec_gpspi_data * const data = dev->data;
+	struct xec_gpspi_regs * const regs = cfg->regs;
+	struct spi_context *ctx = &data->ctx;
+	size_t cur_xfer_len;
+	int counter, ret;
+	uint8_t txb, rxb;
+
+	if ((regs->status & XEC_GPSPI_STS_RXBF_TXBE) != BIT(XEC_GPSPI_STATUS_TXBE_POS)) {
+		return -EBUSY;
+	}
+
+	while (spi_context_tx_buf_on(ctx) || spi_context_rx_buf_on(ctx)) {
+		cur_xfer_len = spi_context_longest_current_buf(ctx);
+
+		for (size_t i = 0; i < cur_xfer_len; i++) {
+
+			/* Write byte to generate SPI clocks */
+			if (spi_context_tx_buf_on(ctx)) {
+				regs->control |= BIT(XEC_GPSPI_CTRL_BI_DIR_OUT_EN_POS);
+				txb = *ctx->tx_buf;
+				regs->tx_data = txb;
+				spi_context_update_tx(ctx, 1, 1);
+			} else {
+				regs->control &= ~BIT(XEC_GPSPI_CTRL_BI_DIR_OUT_EN_POS);
+				regs->tx_data = 0;
+			}
+
+			/* Wait for rx data to fill with one data byte */
+			counter = 0;
+			while ((regs->status & XEC_GPSPI_STS_RXBF_TXBE)
+				!= XEC_GPSPI_STS_RXBF_TXBE) {
+
+				ret = xec_gpspi_spin_yield(&counter, XEC_GPSPI_WAIT_LOOPS);
+				if (ret) {
+					return ret;
+				}
+			}
+
+			/* Get received byte */
+			rxb = regs->rx_data;
+
+			/* Store received byte if rx buffer is on */
+			if (spi_context_rx_on(ctx)) {
+				*ctx->rx_buf = rxb;
+				spi_context_update_rx(ctx, 1, 1);
+			}
+		}
+	}
+
+	spi_context_complete(ctx, dev, 0);
+
+	return 0;
+}
+
+#ifdef CONFIG_SPI_ASYNC
+/* TODO GPSPI controller is byte oriented. DMA can only be done using
+ * the XEC legacy DMA controller (no Zephyr driver). Therefore we
+ * implement asynchronous mode using GPSPI RXBF interrupt.
```

Comment:
Looks like you are mixing up software asynchronicity and interrupt based calls.
If you controller is able to run on interrupt mode, then BOTH non-async and async calls can run on interrupt mode. 

So you should not need any xec_gpspi_xfr_sync() function after all.

Reply:
@tbursztyka, do you agree? or want to implement Sync call using interrupt method?

Reply:
@albertofloyd can you help to provide your feedback on this?